client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,236 @@
+     * @param scopeName the name of the scope tied to the segment, for which a delegation token is to be obtained
+     * @param streamName the name of the stream tied to the segment, for which a delegation token is to be obtained
+     */
+    public JwtTokenProviderImpl(String token, Controller controllerClient, String scopeName,
+                                String streamName) {
+        this(token, controllerClient, scopeName, streamName, ConfigurationOptionsExtractor.extractInt(
+                "pravega.client.auth.token-refresh.threshold",
+                "pravega_client_auth_token-refresh.threshold",
+                DEFAULT_REFRESH_THRESHOLD));
+    }
+
+    /**
+     * Constructs a new object with the specified {delegationToken}, {@code controllerClient}, {@code scopeName} and
+     * {@code streamName}.
+     *
+     * @param token the initial delegation token
+     * @param controllerClient the {@link Controller} client used for obtaining a delegation token from the Controller
+     * @param scopeName the name of the scope tied to the segment, for which a delegation token is to be obtained
+     * @param streamName the name of the stream tied to the segment, for which a delegation token is to be obtained
+     * @param tokenRefreshThreshold the time in seconds before expiry that should trigger a token refresh
+     */
+    public JwtTokenProviderImpl(String token, Controller controllerClient, String scopeName, String streamName,
+                                    int tokenRefreshThreshold) {
+        Exceptions.checkNotNullOrEmpty(token, "delegationToken");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+        this.tokenRefreshThreshold = tokenRefreshThreshold;
+    }
+
+    /**
+     * Returns the delegation token. It returns existing delegation token if it is not close to expiry. If the token
+     * is close to expiry, it obtains a new delegation token and returns that one instead.
+     *
+     * @return String the delegation token JWT compact value
+     */
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public String refreshToken() {
+        log.trace("Entered refreshToken().");
+        DelegationToken existingToken = this.getCurrentToken();
+        if (isTokenBeingRefreshed.get()) {
+            log.debug("Token is already under refresh.");
```

Comment:
When you log these messages, it will help a lot to include context in them. Take a look at some of the classes inside the segment store and how they do it (i.e., `StreamSegmentContainer.java`, `SegmentAggregator.java`, etc.). You should probably include the scope/stream/segment name in all these messages.

Reply:
Sure, I've modified the log statements to include additional context.